### PR TITLE
Allow personal statement to be saved even if over word limit

### PR DESCRIPTION
**Pull request by frankieroberto**
One of the most common errors shown within our service is when people submit a personal statement that is over the 600 word limit.

This happens even though we have a feature which displays the number of words left as you type, and turns the border of the field red and displays an red message if you are over the word limit.

When you submit an answer that is over the word limit, you're returned to the same page with an error message at the top. The answer is not saved until you edit it down to under the word limit.

Our assumption is that either users are not seeing the as-you-type word count, or they are seeing it but want to save anyway and edit it down later.

This PR explores how we might allow users to save a personal statement even if it is over the word limit, but still not allow them to submit their application until it has been edited to within the word limit.

Other minor changes:

* The review page no longer shows "Why you want to be a teacher" within the summary list - this repeats the h1 and uses up extra horizontal space.
* Error message has been reworded from "Tell us why you want to be a teacher must be 600 words or fewer" to "Your answer must be 600 words or less
* Change link text changed from "Change" to "Edit your answer"

🗂️ [Trello card](https://trello.com/c/3M0Cf6MP/1160-explore-if-the-word-count-for-personal-statement-is-higher-than-given-in-word-and-how-we-might-mitigate-this)
➡️ Related: [Update personal statement flow and return to application if empty](https://github.com/DFE-Digital/apply-for-teacher-training/pull/8019)

## How it works

When pressing the green "Continue" button on the personal statement form, the answer is always saved and the user is redirected to the review page.

On the review page, if the answer is above the word count, then an error message is shown, and the user cannot mark the section as completed.

If the answer is below the word count, then the section can be marked as completed.

If a user marks the section as completed, but then edits their answer to be above the word count, then the section is marked as incomplete again.

### Before: Error shown above the form itself

before-error-page

### Before: Section review page

before-review-page

### After: Section review page showing an error

screenshot-localhost_3000-2023 04 20-09_31_27

### After: Section review page showing no errors (mobile view)

localhost_3000_candidate_application_personal-statement_review(iPhone 6_7_8 Plus)

## Notes for reviewers

Firstly, this ended up requiring a lot more code changes than I expected, so please review the code carefully and let me know if you spot any mistakes, confusing code, or opportunities for improvements.

Please also check it in browser! This pattern has been applied to _both_ the old (2 personal statement sections) and new (combined personal statement) flows, so you'll need to check both by turning the feature flag on and off.

Also worth checking on mobile viewports, and seeing if there are any edge cases or flows I'd not thought of.

### CODE REVIEW

**Review by MaeveRoseveare — COMMENTED** [review]
I like it, I like it! A few things:

1. Could we say ‘Edit your statement’ in the link and ‘Your statement must be 600 words or less’ - just thinking of the work we did this morning and that would line up with wording we will use for the merged personal statement

2. I think we should have a ‘Continue’ button at the bottom or it feels like you’re sort of stuck otherwise?

3. Could we use the error summary pattern (so the big red box) to make the error stand out more?

4. Probably a separate thing but the error when you go over the word limit reads a bit oddly, could we say “You're over the word limit by 70 words” or “You're 70 words over the word limit” 
Screenshot 2023-03-10 at 14 38 00

**Comment by MaeveRoseveare** [comment]
@frankieroberto ah ok, so basically I'm still learning about the intricacies of the design system!
For no. 2, I'd expect that Continue button to go back to the task list page. At least that's what makes the most sense to me.

3. I see, I see. I wonder if we take a similar approach then to what we did with the 'combined' personal statement message to say they don't need to change anything. So like: 'Your statement must be 600 words or less. You can edit your statement now, or come back later.' or 'Your statement must be 600 words or less. You do not have to edit it now, but you must edit it before you submit your application.'

4. Ah ok, maybe they have a good reason for it to be like that then. Maybe mine are a bit wordy but maybe slightly clearer?...

**Comment by saliceti** [comment]
Closing to fix review app

**Review by MylesJarvis — APPROVED** [review]
I've tested with and without new personal statement feature flag on and off, appears to work well for me, including carrying over from the old 2 section world, to 1 section.

**Comment by MaeveRoseveare** [comment]
This looks good to me and is working on both merged statement and where we still have the 2 questions. Works on mobile too. Originally I suggested to say 'Edit your statement' instead of 'answer' however, where we still have 2 questions, I don't think 'statement' quite makes sense. So happy to leave that as 'answer' for everything.

File: app/components/candidate_interface/becoming_a_teacher_review_component.html.erb [hunk 1]
```diff
@@ -8,5 +8,18 @@
...
+      </div>
+    <% end %>
+    <div class="app-summary-card__body">
+      <div class="govuk-grid-row">
+        <div class="govuk-grid-column-three-quarters-from-desktop">
+          <%= simple_format(@application_form.becoming_a_teacher, class: 'govuk-body') %>
+        </div>
+      </div>
+    </div>
+  </div>
```
Comment (stevehook) [inline app/components/candidate_interface/becoming_a_teacher_review_component.html.erb:24]: Is there a limitation of the `SummaryCardComponent` that you had to work around here or is it just a completely different layout? (Just curious).

File: app/controllers/candidate_interface/personal_statement_controller.rb [hunk 2]
```diff
@@ -37,16 +40,16 @@ def update
...
+      @becoming_a_teacher_form.save(current_application)
+
+      if @becoming_a_teacher_form.invalid?
+        set_section_to_incomplete_if_completed
+      end
+
+      if @becoming_a_teacher_form.blank?
+        redirect_to candidate_interface_application_form_path
       else
-        track_validation_error(@becoming_a_teacher_form)
```
Comment (stevehook) [inline app/controllers/candidate_interface/personal_statement_controller.rb:48]: Do we lose visibility of other validation errors if we remove this?

File: app/views/candidate_interface/subject_knowledge/show.html.erb [hunk 3]
```diff
@@ -26,13 +26,27 @@
     <% end %>
   <% end %>
 
-  <%= render CandidateInterface::SubjectKnowledgeReviewComponent.new(application_form: @application_form) %>
+  <% if !@subject_knowledge_form.valid? %>
+    <div class="govuk-grid-row">
+      <div class="govuk-grid-column-two-thirds">
+        <div class="govuk-inset-text app-inset-text--error govuk-!-margin-top-0">
+          <h2 class="app-inset-text__title">There is a problem</h2>
+          <p class="govuk-body"><%= @subject_knowledge_form.errors.messages.first[1][0] %></p>
```
Comment (stevehook) [inline app/views/candidate_interface/subject_knowledge/show.html.erb:34]: What happens if there is more than one validation error? Is it possible to use `f.govuk_error_summary`?

**Review by stevehook — COMMENTED** [review]
Tested locally and seems to work perfectly for me. I've left a couple of minor questions inline...

File: app/components/candidate_interface/becoming_a_teacher_review_component.html.erb (hunk 1 above)
Comment (frankieroberto) [reply app/components/candidate_interface/becoming_a_teacher_review_component.html.erb:24]: Yep this version of the summary card doesn’t actually contain a summary list - as in this instance there’s no need for a “Personal statement” label in the first column because there’s only 1 question summarised and that’s the h1 anyway.

I’m not sure if it’s better to extend the summary card component to allow this variant someone, or treat this as a one-off?

File: app/controllers/candidate_interface/personal_statement_controller.rb (hunk 2 above)
Comment (frankieroberto) [reply app/controllers/candidate_interface/personal_statement_controller.rb:48]: Yeah we do. Arguably that’s ok here as this isn’t a traditional validation error as it’s non-blocking? But possibly it’s still useful to track? cc @MylesJarvis

File: app/views/candidate_interface/subject_knowledge/show.html.erb (hunk 3 above)
Comment (frankieroberto) [reply app/views/candidate_interface/subject_knowledge/show.html.erb:34]: I couldn’t use `govuk_error_summary` as that creates links to the fields, expecting them to be on the same page. But in this instance it’s on a different page, and we don’t need links as that’s already on the page below.

In theory there should only ever be 1 error as there’s only 1 field involved - but maybe the code should still display them all?

Comment (stevehook) [reply app/views/candidate_interface/subject_knowledge/show.html.erb:34]: Makes sense. I guess we might have another kind of validation in the future but I don't think we should hold this up.

File: app/components/candidate_interface/becoming_a_teacher_review_component.html.erb (hunk 1 above)
Comment (stevehook) [reply app/components/candidate_interface/becoming_a_teacher_review_component.html.erb:24]: :+1: Sounds like a one-off.

**Review by stevehook — APPROVED** [review]
